### PR TITLE
Add a std::optional and std::variant implementation

### PR DESCRIPTION
**Pull request by leoetlino**
std::optional makes a few things a bit neater and less error prone. However, we still cannot use C++17 (unfortunately), so this commit adds an implementation of std::optional that we can use right now.

To do:
* [x] Fix building with compilers that have C++17 support (e.g. gcc 7.1). The c++17 flag needs to be enabled only when possible
* [x] Fix build failures on gcc 5 when using a unique_ptr with a std::optional. It looks like `std::optional<std::unique_ptr>` is considered to be copy constructible, when it really isn't...

### CODE REVIEW

**Comment by sepalani** [comment]
Should this file be modified or be compliant with our coding style? If so, should we keep these ```== true```, ```== false```?

Otherwise, it isn't used anywhere so I don't think it can harm, for the moment.

**Comment by leoetlino** [comment]
That's pretty easy to change, so there you go. Fixed.

File: Source/Core/Common/StdOptional.h [hunk 1]
```diff
@@ -0,0 +1,913 @@
...
+      clear();
+    else if (!initialized() && rhs.initialized())
+      initialize(*rhs);
+    else if (initialized() && rhs.initialized())
+      contained_val() = *rhs;
+    return *this;
+  }
+
+  optional& operator=(optional&& rhs) noexcept(
+      is_nothrow_move_assignable<T>::value&& is_nothrow_move_constructible<T>::value)
```
Comment (sepalani) [inline Source/Core/Common/StdOptional.h]: There should be a space (after value) between these two booleans.

File: Source/Core/Common/StdOptional.h [hunk 2]
```diff
@@ -0,0 +1,913 @@
...
+  template <class U, class... Args>
+  void emplace(initializer_list<U> il, Args&&... args)
+  {
+    clear();
+    initialize<U, Args...>(il, std::forward<Args>(args)...);
+  }
+
+  // 20.5.4.4, Swap
+  void swap(optional<T>& rhs) noexcept(
+      is_nothrow_move_constructible<T>::value&& noexcept(swap(declval<T&>(), declval<T&>())))
```
Comment (sepalani) [inline Source/Core/Common/StdOptional.h]: Ditto?

File: Source/Core/Common/StdOptional.h (hunk 1 above)
Comment (leoetlino) [inline Source/Core/Common/StdOptional.h]: This is what clang-format wanted to format it to (looks like a bug). I have replaced these ::value with (), which incidentally fixes the missing space issue.

**Comment by Orphis** [comment]
Have you considered calling the file just "optional" and use `__has_include_next` and `#include_next` to use upstream's version instead?
It's supported by Clang and GCC.

**Comment by leoetlino** [comment]
I'm not really familiar with the process of "backporting" newer C++ stuff, so I just followed #351.

Is there any reason for using `#include_next` instead? (And what would its usage look like?) It doesn't seem to be standard or supported by MSVC.

**Comment by Orphis** [comment]
It's not standard anyway to provide anything in the `std` namespace yourself.
`__has_include_next` and `#include_next` are just ways to shadow headers with another one. You can check if there is another `optional` header available, and then use it. Your current file would be called `optional` too. That means the day we want to use the upstream version, we can just remove it.

As for MSVC, it does have it from 2017. We could up the requirement. And there might be a way to conditionally add the include path in vcxproj files.

Another way to do it is to do it is to test if the header exists from CMake. If it does, then we use it, or we add an include path that makes this version available.

Basically, I like compatibility headers to be transparent and with the same name :)

**Comment by Orphis** [comment]
Oh, all of that are suggestions that I think are worth exploring, not requirements in any way.

**Comment by BhaaLseN** [comment]
How else would you use soon-to-be-`std` functionality without putting your own stuff into `std` (sans a ton of `#if standard use this #else ise that`), if not by putting it into a `std` namespace (even when its not something you should be doing normally)?

We've done it before, it worked; but we mostly didn't have other alternatives available. I'm for exploring the alternatives and using standard headers if we can.
Not sure what the general consensus about VS2017 is though (which does have `std::optional` if i read [the docs](https://docs.microsoft.com/en-us/cpp/visual-cpp-language-conformance#standard-library--stl-features) correctly)

**Comment by lioncash** [comment]
IMO, I think updating to VS2017 would be nice.

**Comment by goldenercobra** [comment]
Is this prefferable to using this: [STX Headers](https://github.com/tcbrindle/cpp17_headers)?

**Comment by leoetlino** [comment]
This is pretty much the implementation used in that repo.

Since the move to VS2017 doesn't seem to be happening soon (because of buildbot stuff), how should we proceed with this PR?

**Comment by leoetlino** [comment]
Okay, so we have now moved to VS2017, but I'm not sure about using `#include_next`. With clang, it is only supposed to be used in headers:

>Note that __has_include_next, like the GNU extension `#include_next` directive, is intended for use in headers only, and will issue a warning if used in the top-level compilation file. A warning will also be issued if an absolute path is used in the file argument.

And to be honest, I don't see the benefit over a regular include. Whenever we want to use the upstream version, we would want to remove this implementation *and* replace the `#include_next` with an `#include`, which is not really much more different from changing the include from `Common/StdOptional.h` to `optional`.

**Comment by Orphis** [comment]
You're supposed to use the `#include_next` in the optional header itself, not in the code that uses it.
Another way, is to check in CMake if the header is supported by the compiler and use it. If it's not, then add the header to the include path.

**Comment by BhaaLseN** [comment]
VS 2017 supports `std::optional`, so do we still need this PR?

**Comment by leoetlino** [comment]
@Orphis Oh whoops, looks like I misread your suggestion. I'll change it to use the upstream version where possible with `#include_next`.

@BhaaLseN We do, if we don't want to require gcc 7 everywhere else :P

**Comment by leoetlino** [comment]
@Orphis It's been a while since I opened the PR, so I kind of forgot that we already use the upstream version where possible (`__has_include(<optional>)`). Is that good enough for you? `__has_include` is standard and is part of C++17.

**Comment by Orphis** [comment]
Well, I like compatibility changes to be non intrusive for the users.
So the header that people should include be `<optional>`.

So 2 options:

- You have your header called optional always in the path and you check `__has_include_next(<optional>)` to use upstream version automatically when available, or fallback to your implementation.
- You check from CMake if the compiler supports `<optional>`, if it doesn't you add your compatibility header to the include path.

Is that clear enough now? :)

**Comment by leoetlino** [comment]
@Orphis okay, I think I got it now. Would you mind reviewing this?

File: Source/VSProps/Base.props [hunk 3]
```diff
@@ -80,6 +80,8 @@
       <RuntimeTypeInfo>false</RuntimeTypeInfo>
       <MinimalRebuild>false</MinimalRebuild>
       <MultiProcessorCompilation>true</MultiProcessorCompilation>
+      <!--Enable latest C++ standard-->
+      <LanguageStandard>stdcpplatest</LanguageStandard>
```
Comment (Orphis) [inline Source/VSProps/Base.props]: You should add the same to the CMake files for Windows too.

Comment (leoetlino) [inline Source/VSProps/Base.props]: This was simply cherry-picked from #5484, so it'd have to be fixed there

**Comment by Orphis** [comment]
Looks fine. Let's wait for #5484 first I think!

File: CMakeLists.txt [hunk 4]
```diff
@@ -540,6 +540,13 @@ if(ANDROID)
   include_directories(Source/Android)
 endif()
 
+if(NOT CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
```
Comment (ligfx) [inline CMakeLists.txt]: - Is the intention to expose this for all externals?
- Does this need to be `BEFORE`, since it tries to delegate to the system headers anyways?

Comment (leoetlino) [inline CMakeLists.txt]: Hmm, this should probably only be for our own code, similar to how C++14/C++17 is only enabled for Source/. I'll change this.

File: Source/CMakeLists.txt [hunk 5]
```diff
@@ -13,11 +9,31 @@ if(CMAKE_SYSTEM_NAME MATCHES "Windows")
...
+if(CMAKE_C_COMPILER_ID MATCHES "MSVC")
+  # enable the latest C++ standard feature set,
+  # and also disable MSVC specific extensions
+  # to be even more standards compliant.
+  check_and_add_flag(CPPLATEST /std:c++latest)
+  check_and_add_flag(STANDARD_COMPLIANCE /permissive-)
+else()
+  # Enable C++17, but fall back to C++14 if it isn't available.
+  # This would have been simpler if cmake supported CMAKE_CXX_STANDARD = 17;
+  # it currently doesn't, so this check has to be done manually.
```
Comment (ligfx) [inline Source/CMakeLists.txt]: For CMake version 3.8 and up, `CMAKE_CXX_STANDARD` should be used, to correctly handle cases like https://gitlab.kitware.com/cmake/cmake/issues/16468

It's unfortunate they raise a hard error on an unsupported standard, rather than decaying to a previous one :/

**Comment by ligfx** [comment]
This broke compiling DolphinQt2 with CMake, for the reasons outlined in https://gitlab.kitware.com/cmake/cmake/issues/16468
